injection/clients.go

```diff
@@ -0,0 +1,47 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package injection
+
+import (
+	"context"
+	"sync"
+
+	"k8s.io/client-go/rest"
+)
+
+// ClientInjector holds the type of a callback that attaches a particular
+// client type to a context.
+type ClientInjector func(context.Context, *rest.Config) context.Context
+
+var (
+	clientMutex sync.Mutex
+	clients     []ClientInjector
+)
+
+func RegisterClient(ci ClientInjector) {
+	clientMutex.Lock()
+	defer clientMutex.Unlock()
+
+	clients = append(clients, ci)
+}
+
+func GetClients() []ClientInjector {
+	clientMutex.Lock()
+	defer clientMutex.Unlock()
+
+	return clients[:]
```

Comment:
My understanding was that this was the simplest way to copy it so that it couldn't be modified by an `append()`.  That may be incorrect (and some simple Googling doesn't suggest I am right).

Reply:
It will be a copy, yes.

Reply:
Slices are value types that point to a backing array on a heap.

Returning `clients` is already returning a copy ;)

Reply:
If the intent is to prevent side-effects (ie. replacing an element at some index affects all arrays) then you need to use `make` and copy the elements into the new slice.

Reply:
I guess, we're looking for `append(a[:0:0], a...) ` :-)